src/chat.rs

```diff
@@ -227,23 +215,90 @@ impl ChatId {
-            return false;
+            bail!("ignoring setting of Block-status for {}", self);
         }
-        context
+        let count = context
             .sql
             .execute(
-                "UPDATE chats SET blocked=? WHERE id=?;",
+                "UPDATE chats SET blocked=?1 WHERE id=?2 AND blocked != ?1",
                 paramsv![new_blocked, self],
             )
-            .await
-            .is_ok()
+            .await?;
+        Ok(count > 0)
+    }
+
+    /// Blocks the chat as a result of explicit user action.
+    pub async fn block(self, context: &Context) -> Result<()> {
+        let chat = Chat::load_from_db(context, self).await?;
+
+        match chat.typ {
+            Chattype::Undefined => bail!("Can't block chat of undefined chattype"),
+            Chattype::Single => {
+                for contact_id in get_chat_contacts(context, self).await? {
+                    if contact_id != DC_CONTACT_ID_SELF {
+                        info!(
+                            context,
+                            "Blocking the contact {} to block 1:1 chat", contact_id
+                        );
+                        Contact::block(context, contact_id).await?;
+                    }
+                }
+            }
+            Chattype::Group => {
+                info!(context, "Can't block groups yet, deleting the chat");
+                self.delete(context).await?;
+            }
+            Chattype::Mailinglist => {
+                if self.set_blocked(context, Blocked::Manually).await? {
+                    context.emit_event(EventType::ChatModified(self));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Unblocks the chat.
+    pub async fn unblock(self, context: &Context) -> Result<()> {
+        self.set_blocked(context, Blocked::Not).await?;
+        Ok(())
     }
 
-    pub async fn unblock(self, context: &Context) {
-        self.set_blocked(context, Blocked::Not).await;
+    /// Accept the contact request.
+    ///
+    /// Unblocks the chat and scales up origin of contacts.
+    pub async fn accept(self, context: &Context) -> Result<()> {
+        let chat = Chat::load_from_db(context, self).await?;
+
+        match chat.typ {
+            Chattype::Undefined => bail!("Can't accept chat of undefined chattype"),
+            Chattype::Single | Chattype::Group => {
+                // User has "created a chat" with all these contacts.
+                //
+                // Previously accepting a chat literally created a chat because unaccepted chats
+                // went to "contact requests" list rather than normal chatlist.
+                for contact_id in get_chat_contacts(context, self).await? {
+                    if contact_id != DC_CONTACT_ID_SELF {
+                        Contact::scaleup_origin_by_id(context, contact_id, Origin::CreateChat)
+                            .await;
+                    }
+                }
```

Comment:
This will make the contact visible in your contact list.

I am not sure about groups, but the contacts will become visible in the contact list once you send message to the group. If we want to separate manually added contacts and collected contacts and prevent polluting the contact list with too much random contacts from public groups, all the places that manipulate origin should be revised, it's not a simple change for this PR.